File: make_folder_icon.pl
# This script handles installing and uninstalling custom folder icons on Windows.
# It requires ImageMagick for the 'install' command's image conversion.
#
# Usage:
#   perl script.pl install "C:\My Documents" "C:\Photos\photo.jpg"
#   perl script.pl install "C:\My Documents"
#   perl script.pl uninstall "C:\My Documents"

use strict;
use warnings;
use File::Spec;
use File::Basename;
use Win32 qw();
use File::Glob qw(:bsd_glob);

# Main program logic
my $command = shift @ARGV;

if (!defined $command) {
    die "Error: No command specified. Usage: perl $0 [install|uninstall] <arguments>\n";
}

if ($command eq 'install') {
    install_icon(@ARGV);
} elsif ($command eq 'uninstall') {
    uninstall_icon(@ARGV);
} else {
    die "Error: Unknown command '$command'. Use 'install' or 'uninstall'.\n";
}

exit;

#-----------------------------------------------------------------------------
# Subroutines
#-----------------------------------------------------------------------------

sub install_icon {
    my ($folder_path, $image_path) = @_;

    # Normalize folder path
    $folder_path =~ s|\\|/|g;

    unless (-d $folder_path) {
        die "Error: Folder '$folder_path' does not exist.\n";
    }

    # If only one argument, assume the image is in the folder
    if (@_ == 1) {
        my @found_files = bsd_glob(File::Spec->catfile($folder_path, "folder.*"));
        
        # Filter for common image extensions
        my @valid_images = grep { /\.(png|jpe?g|ico)$/i } @found_files;
        
        if (@valid_images) {
            $image_path = $valid_images[0];
            print "Found image '$image_path' in the folder.\n";
        } else {
            die "Error: No 'folder.png', 'folder.jpg', or 'folder.ico' found in '$folder_path'.\n";
        }
    } elsif (@_ != 2) {
        die "Usage: perl $0 install <folder_path> [<image_path>]\n";
    }
    
    # Normalize image path
    $image_path =~ s|\\|/|g;

    unless (-e $image_path) {
        die "Error: Image file '$image_path' does not exist.\n";
    }

    my $ico_path;
    my ($filename, $dirs, $suffix) = fileparse($image_path, qr/\..*/);

    # Convert image to ICO if needed
    if (lc($suffix) ne '.ico') {
        print "Converting $suffix to .ico file using ImageMagick...\n";
        $ico_path = File::Spec->catfile($dirs, "$filename.ico");
        my $convert_command = "convert \"$image_path\" -define icon:auto-resize=256,64,48,32,16 \"$ico_path\"";
        
        my $result = system($convert_command);
        if ($result != 0) {
            die "Error: ImageMagick 'convert' failed. Make sure it's in your PATH.\n";
        }
        print "Conversion successful. New ICO file created at '$ico_path'.\n";
    } else {
        $ico_path = $image_path;
        print "Image is already an ICO file. Skipping conversion.\n";
    }

    # Use short path name
    my $short_ico_path = '';
    eval {
        $short_ico_path = Win32::GetShortPathName($ico_path);
    };
    $short_ico_path = $ico_path unless $short_ico_path;

    # Create the desktop.ini file
    my $desktop_ini_path = File::Spec->catfile($folder_path, 'desktop.ini');
    my $ini_content = "[.ShellClassInfo]\nIconResource=$short_ico_path,0\nInfoTip=Custom folder icon\n";
    open(my $fh, '>', $desktop_ini_path) or die "Could not write desktop.ini: $!";
    print $fh $ini_content;
    close($fh);

    # Set file and folder attributes
    system("attrib +h +s \"$desktop_ini_path\"");
    system("attrib +r \"$folder_path\"");

    print "Success! Custom icon installed for '$folder_path'.\n";
    print "You may need to press F5 in Explorer to refresh.\n";
}

sub uninstall_icon {
    my ($folder_path) = @_;

    if (@_ != 1) {
        die "Usage: perl $0 uninstall <folder_path>\n";
    }

    # Normalize path
    $folder_path =~ s|\\|/|g;

    unless (-d $folder_path) {
        die "Error: Folder '$folder_path' does not exist.\n";
    }

    my $desktop_ini_path = File::Spec->catfile($folder_path, 'desktop.ini');

    # Check if the desktop.ini file exists
    if (-e $desktop_ini_path) {
        print "Removing '$desktop_ini_path'...\n";
        # First, remove the system and hidden attributes so we can delete the file
        system("attrib -s -h \"$desktop_ini_path\"");
        
        # Then, delete the file
        unlink $desktop_ini_path or die "Error deleting '$desktop_ini_path': $!";
        
        print "File removed successfully.\n";
    } else {
        print "No 'desktop.ini' found in '$folder_path'. Nothing to uninstall.\n";
    }

    # Remove the read-only attribute from the folder
    print "Removing read-only attribute from '$folder_path'...\n";
    system("attrib -r \"$folder_path\"");
    print "Read-only attribute removed.\n";
    
    print "Uninstallation complete. Folder icon will revert to default.\n";
    print "You may need to press F5 in Explorer to refresh.\n";
}

File: make_folder_icon.js
// make_folder_icon.js
// Node.js script to install/uninstall custom folder icons on Windows
// Requires ImageMagick's 'convert' command in PATH for image conversion

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function usage() {
  console.error('Usage: node make_folder_icon.js install <folder_path> [<image_path>]');
  console.error('   or: node make_folder_icon.js uninstall <folder_path>');
  process.exit(1);
}

function getShortPathName(winPath) {
  // Use PowerShell to get the short path name
  try {
    const cmd = `powershell -Command "(Get-Item \"${winPath}\").ShortName"`;
    const shortName = execSync(cmd, { encoding: 'utf8' }).trim();
    if (shortName && shortName !== winPath) {
      return path.join(path.dirname(winPath), shortName);
    }
  } catch {}
  return winPath;
}

function installIcon(folderPath, imagePath) {
  folderPath = path.resolve(folderPath);
  if (!fs.existsSync(folderPath) || !fs.statSync(folderPath).isDirectory()) {
    console.error(`Error: Folder '${folderPath}' does not exist.`);
    process.exit(1);
  }

  if (!imagePath) {
    // Look for folder.png, folder.jpg, folder.ico
    const files = fs.readdirSync(folderPath);
    const found = files.find(f => /^folder\.(png|jpe?g|ico)$/i.test(f));
    if (found) {
      imagePath = path.join(folderPath, found);
      console.log(`Found image '${imagePath}' in the folder.`);
    } else {
      console.error(`Error: No 'folder.png', 'folder.jpg', or 'folder.ico' found in '${folderPath}'.`);
      process.exit(1);
    }
  }
  imagePath = path.resolve(imagePath);
  if (!fs.existsSync(imagePath)) {
    console.error(`Error: Image file '${imagePath}' does not exist.`);
    process.exit(1);
  }

  let icoPath = imagePath;
  if (path.extname(imagePath).toLowerCase() !== '.ico') {
    // Convert to .ico using ImageMagick
    icoPath = path.join(path.dirname(imagePath), path.basename(imagePath, path.extname(imagePath)) + '.ico');
    const cmd = `magick "${imagePath}" -define icon:auto-resize=256,64,48,32,16 "${icoPath}"`;
    console.log('Node.js process PATH:', process.env.PATH);
    try {
      console.log('Converting image to .ico using ImageMagick...');
      execSync(cmd, { stdio: 'inherit' });
      console.log(`Conversion successful. New ICO file at '${icoPath}'.`);
    } catch (err) {
      console.error('Error: ImageMagick "magick" failed. Make sure it\'s in your PATH.');
      if (err.stderr) {
        console.error('STDERR:', err.stderr.toString());
      }
      if (err.stdout) {
        console.error('STDOUT:', err.stdout.toString());
      }
      console.error('Full error:', err);
      process.exit(1);
    }
  } else {
    console.log('Image is already an ICO file. Skipping conversion.');
  }

  // Get short path for desktop.ini
  const shortIcoPath = getShortPathName(icoPath);
  const desktopIniPath = path.join(folderPath, 'desktop.ini');
  const iniContent = `[.ShellClassInfo]\nIconResource=${shortIcoPath},0\nInfoTip=Custom folder icon\n`;
  fs.writeFileSync(desktopIniPath, iniContent);

  // Set file and folder attributes
  execSync(`attrib +h +s "${desktopIniPath}"`);
  execSync(`attrib +r "${folderPath}"`);

  console.log(`Success! Custom icon installed for '${folderPath}'.`);
  console.log('You may need to press F5 in Explorer to refresh.');
}

function uninstallIcon(folderPath) {
  folderPath = path.resolve(folderPath);
  if (!fs.existsSync(folderPath) || !fs.statSync(folderPath).isDirectory()) {
    console.error(`Error: Folder '${folderPath}' does not exist.`);
    process.exit(1);
  }
  const desktopIniPath = path.join(folderPath, 'desktop.ini');
  if (fs.existsSync(desktopIniPath)) {
    console.log(`Removing '${desktopIniPath}'...`);
    execSync(`attrib -s -h "${desktopIniPath}"`);
    fs.unlinkSync(desktopIniPath);
    console.log('File removed successfully.');
  } else {
    console.log(`No 'desktop.ini' found in '${folderPath}'. Nothing to uninstall.`);
  }
  console.log(`Removing read-only attribute from '${folderPath}'...`);
  execSync(`attrib -r "${folderPath}"`);
  console.log('Read-only attribute removed.');
  console.log('Uninstallation complete. Folder icon will revert to default.');
  console.log('You may need to press F5 in Explorer to refresh.');
}

// Main
const [,, command, folderPath, imagePath] = process.argv;
if (!command) usage();
if (command === 'install') {
  if (!folderPath) usage();
  installIcon(folderPath, imagePath);
} else if (command === 'uninstall') {
  if (!folderPath) usage();
  uninstallIcon(folderPath);
} else {
  usage();
}
